.github/actions/generate-attestations/src/attestation.ts

```diff
@@ -0,0 +1,60 @@
+import * as types from "./intoto";
+import fs from "fs";
+import path from "path";
+
+export function createStatement(
+  subjects: types.Subject[],
+  type: string,
+  predicate: object
+): types.IntotoStatement {
+  return {
+    _type: types.INTOTO_TYPE,
+    subject: subjects,
+    predicateType: type,
+    predicate,
+  };
+}
+
+export async function writeAttestations(
+  layoutFile: string,
+  predicateType: string,
+  predicateFile: string
+): Promise<Record<string, string>> {
+  // Read SLSA output layout file.
+  const buffer = fs.readFileSync(layoutFile);
+  const layout = JSON.parse(buffer.toString());
+
+  // Read predicate
+  const predicateBuffer = fs.readFileSync(predicateFile);
+  const predicateJson = JSON.parse(predicateBuffer.toString());
+
+  // Iterate through SLSA output layout and create attestations
+  const ret: Record<string, string> = {};
+  for (const att in layout) {
+    if (att !== "version") {
+      // Validate that attestation path is not nested.
+      if (path.dirname(att) !== ".") {
+        throw Error(`attestation filename must not be nested ${att}`);
+      }
+
+      // Validate that attestation is not already written.
+      if (ret[att] !== undefined) {
+        throw Error(`duplicate attestation file ${att}`);
+      }
+
+      // TODO: How to cast directly into types.Subject[]?
+      const subjectJson: types.Subject[] = JSON.parse(
+        JSON.stringify(layout[att])
+      );
+
+      const attestationJSON = createStatement(
```

Comment:
IIUC, the delegated tool is in charge of setting env variables and other parameters from the GH run, correct?

Reply:
Yes! The predicate would be created in a prior step of a trusted job -- but separating this allows me to modularize the predicate creation

(e.g. create v0.2, v1.0)

Reply:
OK. Will we create some guidance about *what* and *how* to record? So that we can be consistent with our own builders?

Reply:
Hmm good point - yes. I could do some predicate validation on some of these fields. You mentioned predicate type checking earlier -- I wasn't sure if you meant some validation that the prediateType input matches the predicate format defined. I thought about adding that in, and if so, I could do make a `validatePredicate(type, predicate)` that checks (1) if the predicate conforms to the Type and (2) if the predicate contains certain expected fields (context, runner material, some source material, a builder ID, to start?)

Reply:
Sg. Maybe just create an issue for now.

Reply:
linked the issue (https://github.com/slsa-framework/slsa-github-generator/issues/1422) in a TODO